gensim/matutils.py

```diff
@@ -667,45 +667,54 @@ def ret_log_normalize_vec(vec, axis=1):
 
 def unitvec(vec, norm='l2'):
     """Scale a vector to unit length.
-
+    
     Parameters
     ----------
     vec : {numpy.ndarray, scipy.sparse, list of (int, float)}
         Input vector in any format
     norm : {'l1', 'l2'}, optional
         Normalization that will be used.
-
+        
     Returns
     -------
     {numpy.ndarray, scipy.sparse, list of (int, float)}
         Normalized vector in same format as `vec`.
-
+        
     Notes
     -----
     Zero-vector will be unchanged.
-
+    
     """
     if norm not in ('l1', 'l2'):
         raise ValueError("'%s' is not a supported norm. Currently supported norms are 'l1' and 'l2'." % norm)
+    
     if scipy.sparse.issparse(vec):
         vec = vec.tocsr()
         if norm == 'l1':
             veclen = np.sum(np.abs(vec.data))
         if norm == 'l2':
             veclen = np.sqrt(np.sum(vec.data ** 2))
         if veclen > 0.0:
-            return vec / veclen
+            if np.issubdtype(vec.dtype, np.int):
+                vec = vec.astype(np.float)
+                return vec / veclen
+            else:
+                vec /= veclen
+                return vec.astype(vec.dtype)
         else:
             return vec
 
     if isinstance(vec, np.ndarray):
-        vec = np.asarray(vec, dtype=float)
```

Comment:
I think that this line is needed especially for `blas_*` calls, please return it